### ISSUE TITLE
develop engine with well chosen positions ?

### ISSUE THREAD

**Issue by tissatussa** [KierenP/Halogen#239]
hi, i just discovered your engine and it's amazingly strong ! I test several different engines and i use special positions (FEN) to see what each engine thinks.  Do you also use this method ?  Maybe to train a Neural Network ?  Are you interested in some of my examples ?

**Comment by KierenP** [comment]
I would be interested to see any positions that Halogen does poorly in or misevaluates. Do you have any of those?

**Comment by tissatussa** [comment]
Here i present some test positions which clearly have a best move.
In many cases most engines do not find the move and don't even consider it, other engines find it only after a (rather) long time.

I guess it's all about "selectivity" .. but i'm not a programmer of any engine.
Like Halogen v9, some engines do not have MultiPV (MPV), but i think this feature is nice when following their evaluation while dp increases.

NOTE: i used Hash 32 Mb and 1 thread .. i let the engines run max 2 minutes and i wrote the depth (dp) they reached. 

---------------------------------------------------------------------------------------

![position-a](https://user-images.githubusercontent.com/1109281/104340479-060c6e80-54f9-11eb-9c91-287721fa04df.png)

r1b1k2r/pp3pp1/4p1n1/q2pP2p/4B1Q1/P1P2N2/2PB1PPP/R4RK1 w kq - 0 1
White played Qxg6 and won this game

see agadmator's Chess Channel - May 5, 2018
"This Game is Just The Best (correspondence game from 2014)"
https://www.youtube.com/watch?v=ZEY8z71Wn2I

most (good) engines evaluate that only 5 moves are relevant, all other moves are very bad

Halogen v9 (dp 32) - no MPV : i tested until dp 38 but Halogen does not find it

Andscacs v0.95123 - no MPV : finds the move as best from the very start
Komodo v13.01 - MPV 9 : finds the move as best, but its evaluation is not convincing
Laser v1.8 beta - MPV 9 : finds the move as best, but its evaluation is not convincing
LC0 v0.26.1 - MPV 9 : finds the move as best at dp 6, after 15 seconds
Rodent IV v0.32 Bosboom personality - MPV 9 : finds the move as best from the very start
RubiChess v1.9 - MPV 9 : finds the move as best from the very start
SF CFish 100121 NNUE - MPV 9 : finds the move as best from the very start
SF Crystal v3.0 141120 NNUE - MPV 9 : finds the move as best from the very start
SF ShashChess v1.51 NNUE - MPV 9 : finds the move as best from the very start
Xiphox v0.6.1 - no MPV : finds the move as best, almost from the very start

---------------------------------------------------------------------------------------

![position-b](https://user-images.githubusercontent.com/1109281/104340534-145a8a80-54f9-11eb-95bc-9588e2353946.png)

r2q1r2/1b2bpkp/n3p1p1/2ppP1P1/p6R/1PN1BQR1/NPP2P1P/4K3 w - - 0 1
here Qf3-f6 should be best move

while testing, only 1 engine finds it (at dp 19 after 4 minutes) :
SF Crystal v3.0 141120 NNUE - MPV 9
(source at https://github.com/jhellis3/Stockfish/tree/mate_finder )

many other engines i tested do NOT find the move (as best) :
Halogen v9 (dp 32) - no MPV

SF CFish 100121 NNUE (dp 25) - MPV 9
SF CiChess v1.8 230819 (dp 24) - MPV 9
SF CorChess v5.1 (dp 25) - MPV 9
SF Honey Xpro (d 26) - MPV 9
SF MateFinder v9 (dp 22) - MPV 9
SF McBrain v9.6 (dp 25) - MPV 9
SF ShashChess v1.51 NNUE (dp 24) - MPV 9
SF Sting v9.99 (dp 14) - MPV 9
SF SugaR v2.6 (dp 21) - MPV 9
SF v12 NNUE - 121020 (dp 26) - MPV 9

Andscacs v0.95123 (dp 20) - no MPV
Arasan v20.1.0 (dp 23) - MPV 9
Ethereal v12.79 (dp 25) - MPV 9
Godel v6.0 (dp 20) - MPV 9
Koivisto v4.1 (dp 29 after 3 minutes) - no MPV - it found Qf3-f6 as first (!) best move but then never considered it again
Komodo v12.1.1 (dp 18) - MPV 9
Komodo v13.01 (dp 20) - MPV 9
Laser v1.8 beta (dp 20) - MPV 9
LC0 v0.26.1 (dp 12 after 7 minutes) - MPV 9 : does consider it : it's second best and not convincing
Minic v3.01 (dp 20) - MPV 9
Nemorio v6.04 (dp 17) - MPV 9
Pedone v2.1 (dp 21) - MPV 9
Rodent IV v0.32 - Tal personality (dp 19) - MPV 9
rofChade v2.3 (dp 25) - no MPV
RubiChess v1.9 (dp 22) - MPV 9
Shredder Classic v5 (dp 21) - MPV 9
Stash v13.0.1 (dp 13) - MPV 9
Texel v1.08a13 (dp 17) - MPV 9
Toga II v4.01 (dp 19) - MPV 9
Wasp v4.0 (dp 22) - MPV 9
Winter v0.9.1 (dp 21) - no MPV
Xiphox v0.6.1 (dp 34) - no MPV

---------------------------------------------------------------------------------------

![position-c](https://user-images.githubusercontent.com/1109281/104340603-263c2d80-54f9-11eb-821d-faefdf5d13f0.png)

invisible to Engines | One Of The Greatest Moves Ever Played
see https://www.youtube.com/watch?v=yGnpewUKP88
"Marshalling his Forces" 
Kramnik - Leko World Championship Match (2004), Brissago SUI, rd 8, Oct-07 

Vladimir Kramnik vs Peter Leko
Spanish Game: Marshall Attack. Modern Main Line (C89)
1. e4 e5 2. Nf3 Nc6 3. Bb5 a6 4. Ba4 Nf6 5. O-O Be7 6. Re1 b5 7. Bb3 O-O 8. c3 d5 9. exd5 Nxd5 10. Nxe5 Nxe5 11. Rxe5 c6 12. d4 Bd6 13. Re1 Qh4 14. g3 Qh3 15. Re4 g5 16. Qf1 Qh5 17. Nd2 Bf5 18. f3 Nf6 19. Re1 Rae8 20. Rxe8 Rxe8 21. a4 Qg6 22. axb5 Bd3 23. Qf2 Re2 24. Qxe2 Bxe2 25. bxa6 Qd3! 26. Kf2 Bxf3 27. Nxf3 Ne4+ 28. Ke1 Nxc3 29. bxc3 Qxc3+ 30. Kf2 Qxa1 31. a7 h6 32. h4 g4

position after 25. bxa6 :
6k1/5p1p/P1pb1nq1/6p1/3P4/1BP2PP1/1P1Nb2P/R1B3K1 b - - 0 1
find Qd3 easily !? .. all engines should find this 'best move' fast .. indeed they do, most find it even at low dp.

Halogen v9 considers Bd6-b8 as best until dp 31, but at greater dp it finds Qg6-d3 very best (+5).
It finds Qg6-d3 as best move from the very start, but changes its evaluation rapidly, until dp > 31.

---------------------------------------------------------------------------------------

![position-d](https://user-images.githubusercontent.com/1109281/104340651-32c08600-54f9-11eb-91f2-5a6b35664c55.png)

r1b2rk1/pp2q1pp/1b6/1BppP3/5Pn1/P1B2RN1/1P2Q1PP/R5K1 w - - 0 1

Ruy Lopez Schliemann Defence Chess Opening: Leela Chess ID 467 vs Stockfish 6 on one cpu
see https://www.youtube.com/watch?v=9rJHxwpVn2Y

Leela (LC0) played Bb5-c4 (!) but my LC0 v0.26.1 does not consider the move at all !?
NOTE: the other main move is h2-h3, considered best by most engines

SF ShashChess v1.51 NNUE - MPV 9 : evaluates the move as best from dp 19
SF Crystal v3.0 141120 NNUE - MPV 9 : finds the move as best from depth 15
SF v12 NNUE - 121020 - MPV 9 : finds the move as best from depth 20
Ethereal v12.79 - MPV 9 : finds the move as best from depth 21

**Comment by KierenP** [comment]
Thank you very much for the hard work you put into this, I'll carefully read through it and analyse with Halogen myself when I get a chance in the next few days and see if I can provide any insight.

**Comment by tissatussa** [comment]
Glad to help .. i thank you for being serious about it .. i'm too .. Halogen seems to be a "new kid on the block", but it's rather strong .. my favorite engine at this moment is ShashChess - find it at GitHub and read its description .. it has 3 personalities (Tal / Capablanca / Petrosian) which can be combined and it has NNUE .. take your time to study my info .. after that i might have some other issues to discuss, while i'm looking to find some (clever) chess programmer who can correspond with me and until now i didn't find anyone .. BTW. my own playing strength is not very high, about 1830, but i'm a chess fanatic all my life .. i'm a computer programmer also - mainly web based. *** bonus: "the Psycho-Cowboy is back" : https://www.bitchute.com/video/g7qApBpHMtN9/ and find more (chess) videos at this channel of mine :-) Good luck !

**Comment by KierenP** [comment]
I've had a close look at all of the positions. They are really tough! Its made me consider adding a multi-PV mode into Halogen.

For analysis below, I used the latest development build of Halogen with 8 cores and 4GB hash.

For position 1, Halogen doesn't consider Qxg6, but at high depths does briefly consider c4. Even after playing Qxg6, fxg6 Halogen doesn't see it as winning. Stockfish only evaluates Qxg6 as somewhere between +2 and +3. I think Halogen is under evaluating the passed pawn that gets created and the lack of black king safety.

For position 2, seeing that Qf6 is winning is extremely hard. Even after playing Qf6, Bxf6 it takes a long time for Halogen to realise the position is winning. After about 30 seconds its had repeated fail highs to realise the score is at least +4.46 but it takes a lot longer to actually finish searching that depth (at least 2 minutes). 

For position 3, Halogen does find Qd3 to be winning so at least it got one right.

For position 4, Halogen seems to want to play h3. After playing Bd4, bxc4, Halogen does eventually see it as winning (+1.35) after a minute or so. Again, pretty tough for Halogen to spot that the sacrifice eventually becomes winning.

**Comment by tissatussa** [comment]
> I've had a close look at all of the positions. They are really tough! Its made me consider adding a multi-PV mode into Halogen.


thank you! So you were that eager :-)  as I stated, MPV is really nice while analysing a position .. some positions have multiple best moves because their evaluation value is almost equal and differs only slightly with increasing dp .. while comparing engines, MPV gives better insight of the evaluation process .. I know positions where the (clearly) best move is ONLY found when setting MPV rather high, eg. 20 (!), otherwise that best move is NEVER considered while dp increases and thus will never end up as best eventually.

when using CuteChess, I do set MPV 1 because this seems to be best practice .. lately I even had some discussion about this somewhere on GitHub and 'experts' pointed me out I made an error by setting it to 5 or 9 during engine-vs-engine play .. I hope Halogen will have MPV in the near future.


> For analysis below, I used the latest development build of Halogen with 8 cores and 4GB hash.


I still have no clue how to use Hash .. engines have (very) different default values for it, maybe because their architecture differs that much .. I read that setting Hash to high is not good practice, because using much Hash can slow down the evaluation instead of helping it. The Readme text of some engine even pointed out a method for the user to determine an optimal Hash value for their PC, but I don't remember which engine it was, sorry .. I used that procedure (for that engine) but the results were unclear to me .. I think 4 GB is way to much, but I can be wrong, not being a chess engine programmer myself.


> For position 1, Halogen doesn't consider Qxg6, but at high depths does briefly consider c4. Even after playing Qxg6, fxg6 Halogen doesn't see it as winning. Stockfish only evaluates Qxg6 as somewhere between +2 and +3. I think Halogen is under evaluating the passed pawn that gets created and the lack of black king safety.


SF (12) indeed evaluates it around +2 but here it's important that SF thinks Qxg6 clearly is best move and all other moves are much worse .. I think it's impressive to see many engines consider Qxg6 best move from the very start : I learned that engines normally hold on to the best move they find first and then try to prove / find another move is better .. I also read ELO is officially determined by doing tests in "micro seconds" : to me it seems their evaluation algorithms are compared this way - correct me if I'm wrong .. you could take a closer look at the latest Rodent IV engine (v0.32) .. this project has now ended, the creator decided he no longer has time to develop it .. Rodent is mainly focused on 'personalities', not (!) playing strength : each personality has different parameters and the user can tweak those .. especially the Bosboom personality (an extarordinary Dutch player) is interesting, it considers Qxg6 from start.


> For position 2, seeing that Qf6 is winning is extremely hard. Even after playing Qf6, Bxf6 it takes a long time for Halogen to realise the position is winning. After about 30 seconds its had repeated fail highs to realise the score is at least +4.46 but it takes a lot longer to actually finish searching that depth (at least 2 minutes).


Yes, this position is tough, only one of my tested engines finds it, using MPV 9 : SF Crystal v3.0, after 4 minutes at dp 19 it gives eval +6.5 and all other moves are -1 or less .. I think selectivity is important here .. today i tested extensively : when MPV is 'low', I mean even lower then 30 (!), many engines do not consider Qf6+ at all and will never consider it when dp increases .. but Qf6+ Kg8 is mate in 9, so Black should accept the Queen sac : Qf6+ Bxf6 gxf6+ and then Kg8 is mate in 8 and Kh8 is mate in 6, so Black should then also offer the Queen to survive, but that way gave the Bishop for 2 pawns and is lost (eval -7 or so). Eg. Ethereal (v12.79) needs MPV 50 to even 'consider' Qf6+ and finds it very best at dp 28 with eval +6.1 ..


> For position 3, Halogen does find Qd3 to be winning so at least it got one right.


Using my own GUI (made with TKinter / Python) I see that Halogen v9 thinks Qd3 is best upto dp 3 (eval +3.23) but then changes its mind : for a long time Bb8 is best with eval about 0.10 .. eventually it finds Qd3 best .. I used the following test settings (and 2 threads) :

with hash   2 Gb it finds Qd3 at dp 36, eval +0.44 but then +1.76
with hash 128 Mb it finds Qd3 at dp 38, eval +0.44 but then +1.76
with hash  32 Mb it finds Qd3 at dp 40, eval +0.44 but then +1.76 
with hash  16 Mb it finds Qd3 at dp 39, eval +0.44 but then +1.76
with hash   4 Mb it does not find Qd3 (I went upto dp 43 - it took about 12 minutes)

Halogen v9 calculates very rapidly : it takes about 90 seconds to reach dp 35 but then it takes a long time going to greater dp.


> For position 4, Halogen seems to want to play h3. After playing Bd4, bxc4, Halogen does eventually see it as winning (+1.35) after a minute or so. Again, pretty tough for Halogen to spot that the sacrifice eventually becomes winning.


[ I skip my comments on this one ]

Why does Halogen v9 calculate so rapidly compared to most other engines ?  It seems nice, you could have made effort to implement a clever move selectivity, but this might also skip some possible moves which could be found best at high dp .. anyway, any solution has its disadvantages, i guess :-)

**Comment by KierenP** [comment]
> when using CuteChess, I do set MPV 1 because this seems to be best practice .. lately I even had some discussion about this somewhere on GitHub and 'experts' pointed me out I made an error by setting it to 5 or 9 during engine-vs-engine play .. I hope Halogen will have MPV in the near future.

Yeah, when playing engines against each other leave the MPV at 1. MPV is only for engine analysis and detrimental for engine strength. 

> I think 4 GB is way to much, but I can be wrong, not being a chess engine programmer myself.

There's no perfect value, typically the more the better. It's true that it can slow down an engine. I presume your GUI shows you how full the hash is? If so, pick a size that doesn't get filled up and only gets to say 50% after the allotted time is up.



> Why does Halogen v9 calculate so rapidly compared to most other engines

I've got no idea. I think my pruning scheme allows for more pruning at lower depths. This has given me some ideas though to test.

**Comment by tissatussa** [comment]
> I presume your GUI shows you how full the hash is? If so, pick a size that doesn't get filled up and only gets to say 50% after the allotted time is up.

Yes, that was the test and the Readme ordered to test this way in the starting position.

> This has given me some ideas though to test.

OK, so I guess I will close this "issue" soon, good luck in building a next Halogen version !  I have some more ideas and questions too, which i might post later as another issue ..

**Comment by KierenP** [comment]
You may be happy to know that Halogen now supports Multi-PV mode, as of (https://github.com/KierenP/Halogen/pull/246). Be sure to clone and build the latest master if you want to use it before I make the official release of Halogen 10 some time in the future.

**Comment by tissatussa** [comment]
Thanks !  I just compiled that version and tested it briefly : all seems to work fine !  MPV is really useful and people can always set MPV 1 to revert to the previous version.  Did you make any other changes, eg. according to my input ?

**Comment by KierenP** [comment]
No other changes yet, but if your work inspires any good ideas in me I’ll
be sure to let you know

**Comment by tissatussa** [comment]
I just discovered a small bug : when setting MPV higher than the amount of legal moves in the position, Halogen stops .. this is logical, but most engines do not have a problem with such setting .. i think this "bug" is easy to fix ..

**Comment by tissatussa** [comment]
you can keep it simple : adjust the "internal" MPV value according to the max number of legal moves.  Eg. when i analyse the 2nd position (where White should find Qf6+) with a GUI, I may have set my MPV to 10 by default, but when I put the Queen on f6 to "help" the engine and see what it finds, we'll have a position where it's Blacks turn and the engine has only two legal moves : Bxf6 or Kg8 .. so you should limit the internal MPV to 2 and your algorithm won't crash ..

**Comment by KierenP** [comment]
Fixed in 9c2b012d1819db1c21a73e2fbce43f312bf86fbe
